### DISCUSSION TITLE
Training spacy v3 NER transformer pipeline on Prodigy annotated dataset

### DISCUSSION THREAD

**Discussion by salman1993** [explosion/spaCy#6929] [Help: Coding & Implementations] [answered]
I have a NER dataset annotated using Prodigy with 3 entity labels: LOCATION, PERSON, DATE

The dataset is in JSONL format with each line looking like this:
```
{"text":"Is there an outage in North York ?","_input_hash":-1484012885,"_task_hash":2032773799,"tokens":[{"text":"Is","start":0,"end":2,"id":0},{"text":"there","start":3,"end":8,"id":1},{"text":"an","start":9,"end":11,"id":2},{"text":"outage","start":12,"end":18,"id":3},{"text":"in","start":19,"end":21,"id":4},{"text":"North","start":22,"end":27,"id":5},{"text":"York","start":28,"end":32,"id":6},{"text":"?","start":33,"end":34,"id":7}],"_session_id":null,"_view_id":"ner_manual","spans":[{"start":22,"end":32,"token_start":5,"token_end":6,"label":"LOCATION"}],"answer":"accept"}
```
I was able run the the [ner_wikiner pipeline](https://github.com/explosion/projects/tree/v3/pipelines/ner_wikiner) from the projects repo. I could also use a transformer NER pipeline from [this repo](https://github.com/explosion/projects/blob/v3/benchmarks/ner_conll03/configs/transformer.cfg) and its working on the WikiNER dataset.

Questions:
1. Whats the best way to convert my Prodigy annotated NER dataset in JSONL format to a corpus that I can save to spacy's binary training format and then train using spacy v3 [transformer, ner] pipeline? This would be the equivalent of what happens when I run `spacy project run corpus` workflow
2. Do I have to convert the dataset to IOB or BILOU format or are there existing scripts to do that from Prodigy files?
3. If I initialize the transformer with pre-trained vectors, does it start off with the Huggingface pretrained weights or does it use weights from spacy's NER model which was trained on the Ontonotes dataset? Just trying to understand whats happening under the hood.
4. What training strategy would be the best for a small training set (~6K samples)? Any pointers here would be much appreciated (i am exploring spacy v3 for first time)?

Versions:
```
spacy              3.0.1
spacy-transformers 1.0.1
prodigy            1.10.4
```

**Comment by ines** [comment, marked as answer]
> Whats the best way to convert my Prodigy annotated NER dataset in JSONL format to a corpus that I can save to spacy's binary training format and then train using spacy v3 [transformer, ner] pipeline? This would be the equivalent of what happens when I run spacy project run corpus workflow

At the moment, the easiest way is to use `data-to-spacy` to export a JSON file, and then convert that to the binary format using `spacy convert`. There'll also be a Prodigy nightly release very soon that uses spaCy v3 and will output `.spacy` files and configs instead (among other things).

> Do I have to convert the dataset to IOB or BILOU format or are there existing scripts to do that from Prodigy files?

Nope, that all happens automatically in `data-to-spacy`.

> If I initialize the transformer with pre-trained vectors, does it start off with the Huggingface pretrained weights or does it use weights from spacy's NER model which was trained on the Ontonotes dataset? Just trying to understand whats happening under the hood.

I'm not 100% sure I understand the question correctly – if you're training a transformer-based pipeline, you'll typically start off with a blank model and the transformer component used for token-to-vector encoding. If you want, you can source some components from existing pipelines, for instance, to keep updating a trained entity recognizer. You can read more about this here: https://spacy.io/usage/training#config-components (In this case, the same token-to-vector encoding component should be available – if you copy an NER component that was trained with transformer weights into a pipeline without transformer weights, it's not going to work.)

> What training strategy would be the best for a small training set (~6K samples)? Any pointers here would be much appreciated (i am exploring spacy v3 for first time)?

After you've converted you data, I'd recommend to just start off with a default config and try one experiment with transformers (and maybe one without, if you want to compare the speed/accuracy trade-offs) 🙂 You shouldn't have to tweak any parameters at this point. You can use the [quickstart generator](https://spacy.io/usage/training#quickstart) or `spacy init config` to auto-generate a starter config for your use case.

**Reply by salman1993** [reply]
Hi @ines, thanks for helping out! Really enjoying spacy v3. I was able to train a NER model with and without transformer and it's working quite well :) 

3 - Initializing Transformer model
I was wondering if I could I initialize the transformer model with the weights from the `en_core_web_trf` model since it has been trained on lot of NER data. My dataset is small so wondering if initializing with those weights could help the model perform better on my dataset.


1,2 - Data conversion
For the data conversion, I used the [preprocessing script](https://github.com/explosion/projects/blob/v3/tutorials/ner_drugs/scripts/preprocess.py) here and it worked (posting for anyone else reading this in the future).

**Reply by ines** [reply]
>  I was wondering if I could I initialize the transformer model with the weights from the en_core_web_trf model since it has been trained on lot of NER data. My dataset is small so wondering if initializing with those weights could help the model perform better on my dataset.

If you're training the same labels, then it can definitely make sense to start with the existing trained component and update it with more data. In your config, you can then "source" the `ner` component from `en_core_web_trf` instead of creating it from scratch: https://spacy.io/usage/training#config-components

Just make sure your data includes enough examples of what the model previously predicted correctly and all the labels you're interested in – you don't want it to overfit on your new small dataset and "forget" anything. You can easily create this extra data, though: just run the model over a bunch of relevant text, let it predict entities and include those examples in your training data. (Pro tip: The binary `.spacy` format is nothing more than a serialized [`DocBin`](https://spacy.io/api/docbin), a collection of `Doc` objects. And `spacy train` accepts directories of multiple `.spacy` files for training. So you can just process some documents with the existing model, save them out and you have additional auto-generated data to "remind" it of its previous predictions ✨)

If the labels you want to train are very different from what the existing `ner` component was trained to predict, it's typically better to start from scratch. Otherwise, you'd be constantly "fighting" the existing weights and trying to teach your model to suddenly predict something completely different from all the data it's seen before. This makes it much harder to reason about the results and can easily lead to confusing behaviour and worse accuracy.

**Reply by salman1993** [reply]
Thanks for the tips - very helpful! 🙂   A couple more questions on creating a larger dataset since I am noticing the model when trained on just my small dataset does show "catastrophic forgetting" as you mentioned.

Can you clarify what you mean by same labels? For my task - there are 3 labels: LOCATION, PERSON, DATE which is a subset of the labels in spacy NER model. They are the same since GPE = LOCATION but we do not contain all the same labels as spacy.

I was hoping to add samples from larger datasets such as WikiNER so that it helps with "forgetting". My domain is a bit different but I am hoping this will help - dev and test are still from my domain only so I can evaluate to see if it helps. If I wanted to add in samples containing LOC and PER from [ner_wikiner](https://github.com/explosion/projects/tree/v3/pipelines/ner_wikiner) dataset - is there a quick way to create dataset using NER mapping? 
```
spacy convert assets/iob corpus --ner-map {"LOC": "LOCATION", "MISC": "O", "ORG": "O", ...}
```

**Reply by ines** [reply]
> Can you clarify what you mean by same labels? For my task - there are 3 labels: LOCATION, PERSON, DATE which is a subset of the labels in spacy NER model. They are the same since GPE = LOCATION but we do not contain all the same labels as spacy.

If it's a subset, it will probably work fine – although, the `GPE` and `LOCATION` distinction is a bit unideal, because you'll now be showing the model a lot of `LOCATION` examples that it previously would have labelled `GPE`. What I would see as problematic is if you only wanted to use one label, e.g. `PERSON`, and then had a bunch of new labels that conflicted with the rest, e.g. custom stuff with numbers that overlaps with `DATE`/`CARDINAL`/`ORDINAL`, or something like that.

> dev and test are still from my domain only so I can evaluate to see if it helps

Ah, this is great 💯 This also sets you up super well for different experimens. For example, what happens if you run the `en_core_web_trf` model over _lots_ of raw, unlabelled text, extract only `PERSON`, `DATE`, `GPE` and `LOCATION`, change the label of all `GPE` entities to `LOCATION` to match your label scheme and then train a new model on this from scratch, evaluated against your in-domain data? So basically, only have the model auto-label data from your domain. What accuracy do you get out of the box? This could be your baseline, and then you can see how far you can improve it by adding more custom domain-specific annotations.

(I'm actually genuinely curious so please report back if you try this!)

> is there a quick way to create dataset using NER mapping?

Doesn't the `convert` command handle this (if you give it a JSON file for the `--ner-map`)?

**Reply by salman1993** [reply]
Hi Ines, following up on this - I tried using 2 things to compare to just training the model on our dataset:
1. Using `en_core_web_trf` to predict on text and use that as part of training -> the results were not as good as just training on our smaller dataset. 
2. Combining some Twitter NER datasets such as BTC and Ritter and using the location and person entities (they do not have date). This did not improve results either. The reason behind using Twitter is that it is somewhat closer to our domain.

I think the main reason its hard to improve on our task is because the domain is quite different (num of tokens, type of text, etc.) so the transfer is not so good. It is helpful to have the transformer initialized with pre-trained vectors but adding in the noisy extra data did not help unfortunately.

Thanks for your help! 😊
